backend/src/main/java/com/redhat/cloud/notifications/routers/internal/errata/ErrataUserPreferencesMigrationResource.java

```diff
@@ -0,0 +1,96 @@
+import com.redhat.cloud.notifications.auth.ConsoleIdentityProvider;
+import io.quarkus.logging.Log;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import jakarta.annotation.security.RolesAllowed;
+import jakarta.inject.Inject;
+import jakarta.validation.constraints.NotNull;
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.core.MediaType;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVRecord;
+import org.jboss.resteasy.reactive.RestForm;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+@Path(Constants.API_INTERNAL + "/team-nado")
+@RolesAllowed(ConsoleIdentityProvider.RBAC_INTERNAL_ADMIN)
+public class ErrataUserPreferencesMigrationResource {
+    private static final String USERNAME_KEY = "username";
+    private static final String ORG_ID_KEY = "org_id";
+    private static final String[] CSV_HEADERS = {USERNAME_KEY, ORG_ID_KEY};
+
+    @Inject
+    ErrataMigrationRepository errataMigrationRepository;
+
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Path("/migrate/csv")
+    @POST
+    public void migrateTeamNadoUserPreferencesCSV(@RestForm("csvFile") InputStream csvFile) throws IOException {
+        final byte[] contents = csvFile.readAllBytes();
+
+        // Set the format for the CSV file.
+        final CSVFormat csvFormat = CSVFormat.DEFAULT
+            .builder()
+            .setHeader(CSV_HEADERS)
+            .setRecordSeparator(System.lineSeparator())
+            .setSkipHeaderRecord(true)
+            .build();
+
+        // Extract the subscriptions from the CSV file.
+        final List<ErrataSubscription> errataSubscriptions = new ArrayList<>();
+        final Reader csvReader = new InputStreamReader(new ByteArrayInputStream(contents));
+        for (final CSVRecord record : csvFormat.parse(csvReader)) {
+            final String username = record.get(USERNAME_KEY);
+            final String orgId = record.get(ORG_ID_KEY);
+
+            errataSubscriptions.add(new ErrataSubscription(username, orgId));
+        }
+
+        Log.infof("Read CSV errata subscriptions file and extracted %s subscriptions from it", errataSubscriptions.size());
+
+        try {
+            this.errataMigrationRepository.saveErrataSubscriptions(errataSubscriptions);
+        } catch (final Exception e) {
+            Log.error("Unable to migrate errata subscriptions due to an exception. The insertions were rolled back", e);
+        }
+    }
+
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Path("/migrate/json")
+    @POST
+    public void migrateTeamNadoUserPreferencesJSON(@NotNull @RestForm("jsonFile") InputStream jsonFile) throws IOException {
+        final byte[] contents = jsonFile.readAllBytes();
```

Comment:
Do we have an estimate of the file size already? Loading all bytes in memory could be an issue if it's too big.

Reply:
No, we could be talking about 20,000 subscriptions, so it might be a few megs of size, I guess. The thing is that if we use a JSON file I don't think we can simply read line by line...

Reply:
Good point. I wrote my comment with the CSV file in mind, but we don't have a choice with JSON.

Reply:
If it helps, we could intentionally break the data dump into multiple files. (It's also possible to stream over the elements of a json array if you have jackson on your class path [[example](https://stackoverflow.com/a/24840488)].)

Reply:
All right, refactored! Now we process the JSON file in batches and we create the corresponding objects at that time, by cleaning them up afterwards so that we don't use as much memory.